### PR TITLE
add vander function

### PR DESCRIPTION
**Pull request by KsanaKozlova**

### CODE REVIEW

File: dpnp/dpnp_iface_arraycreation.py [hunk 1]
```diff
@@ -1122,6 +1123,45 @@ def triu(m, k=0):
...
+           [ 27,   9,   3,   1],
+           [125,  25,   5,   1]])
+    >>> np.vander(x, increasing=True)
+    array([[  1,   1,   1,   1],
+           [  1,   2,   4,   8],
+           [  1,   3,   9,  27],
+           [  1,   5,  25, 125]])
+    """
+    if (not use_origin_backend(x1)):
+        if x1.ndim != 1:
```
Comment (densmirn) [inline dpnp/dpnp_iface_arraycreation.py:1155]: We need to check `x1` is `dparray` here.

File: dpnp/dpnp_iface_arraycreation.py [hunk 2]
```diff
@@ -1122,6 +1123,45 @@ def triu(m, k=0):
...
+            pass
+        else:
+            if N is None:
+                N = len(x1)
```
Comment (densmirn) [inline dpnp/dpnp_iface_arraycreation.py:1159]: Why not `x1.size`?

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 3]
```diff
@@ -127,6 +127,38 @@ void dpnp_ones_like_c(void* result, size_t size)
     dpnp_ones_c<_DataType>(result, size);
 }
 
+template <typename _DataType_input, typename _DataType_output>
+void dpnp_vander_c(void* array1_in, void* result1, size_t size_in, size_t N, int increasing)
```
Comment (densmirn) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:131]: Let's use `const` where it's possible.

File: dpnp/backend/include/dpnp_iface.hpp [hunk 4]
```diff
@@ -843,6 +843,20 @@ template <typename _DataType_input1, typename _DataType_input2, typename _DataTy
...
+ *
+ * @param [in]  array_in    Input array.
+ * @param [out] result      Output array.
+ * @param [in]  size_in     Number of elements in the input array.
+ * @param [in]  N           Number of columns in the output.
+ * @param [in]  increasing  Order of the powers of the columns.
+ * 
+ */
+template <typename _DataType_input, typename _DataType_output>
+INP_DLLEXPORT void dpnp_vander_c(void* array1_in, void* result1, size_t size_in, size_t N, int increasing);
```
Comment (samir-nasibli) [inline dpnp/backend/include/dpnp_iface.hpp:858]:
```suggestion
INP_DLLEXPORT void dpnp_vander_c(const void* array_in, void* result, const size_t size_in, const size_t N, const int increasing);
```

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 5]
```diff
@@ -127,6 +127,38 @@ void dpnp_ones_like_c(void* result, size_t size)
...
+{
+    if (!size_in or !N) return;
+    _DataType_input* array_in = reinterpret_cast<_DataType_input*>(array1_in);
+    _DataType_output* result = reinterpret_cast<_DataType_output*>(result1);
+
+    if (N == 1) {
+        for (size_t i = 0; i < size_in; ++i) { result[i] = 1; }
+        return;
+    }
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:140]: Bad style.
try to run: `python ./setup.py style -a` for auto style.

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 6]
```diff
@@ -127,6 +127,38 @@ void dpnp_ones_like_c(void* result, size_t size)
     dpnp_ones_c<_DataType>(result, size);
 }
 
+template <typename _DataType_input, typename _DataType_output>
+void dpnp_vander_c(void* array1_in, void* result1, size_t size_in, size_t N, int increasing)
+{
+    if (!size_in or !N) return;
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:133]:
```suggestion
    if (!size_in || !N)
    {
        return;
    }
```

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 7]
```diff
@@ -127,6 +127,38 @@ void dpnp_ones_like_c(void* result, size_t size)
...
+    _DataType_input* array_in = reinterpret_cast<_DataType_input*>(array1_in);
+    _DataType_output* result = reinterpret_cast<_DataType_output*>(result1);
+
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:136]: add, please nullptr check.

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 8]
```diff
@@ -394,21 +426,28 @@ void func_map_init_arraycreation(func_map_t& fmap)
...
+    fmap[DPNPFuncName::DPNP_FN_VANDER][eft_BLN][eft_BLN] = {eft_LNG, (void*)dpnp_vander_c<bool, long>};
+    fmap[DPNPFuncName::DPNP_FN_VANDER][eft_C128][eft_C128] = {eft_C128, (void*)dpnp_vander_c<std::complex<double>, std::complex<double>>};
 
     fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_INT][eft_INT] = {eft_INT, (void*)dpnp_zeros_c<int>};
     fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_LNG][eft_LNG] = {eft_LNG, (void*)dpnp_zeros_c<long>};
     fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_FLT][eft_FLT] = {eft_FLT, (void*)dpnp_zeros_c<float>};
     fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_DBL][eft_DBL] = {eft_DBL, (void*)dpnp_zeros_c<double>};
     fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_BLN][eft_BLN] = {eft_BLN, (void*)dpnp_zeros_c<bool>};
-    fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_C128][eft_C128] = {eft_C128, (void*)dpnp_ones_c<std::complex<double>>};
+    fmap[DPNPFuncName::DPNP_FN_ZEROS][eft_C128][eft_C128] = {eft_C128, (void*)dpnp_zeros_c<std::complex<double>>};
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:442]: Not related. It is better to have this fix in separate PR

File: dpnp/dpnp_iface_arraycreation.py [hunk 9]
```diff
@@ -1122,6 +1123,45 @@ def triu(m, k=0):
...
+            return dpnp_vander(x1, N, increasing)
+
+    return numpy.vander(x1, N=N, increasing=increasing)
```
Comment (samir-nasibli) [inline dpnp/dpnp_iface_arraycreation.py:1162]:
```suggestion
    return call_origin(numpy.vander, x1, N=N, increasing=increasing)
```

File: tests/test_arraycreation.py [hunk 10]
```diff
@@ -193,3 +193,29 @@ def test_triu_size_null(k):
...
+                         ids=['float64', 'float32', 'int64', 'int32', 'bool', 'complex128'])
+@pytest.mark.parametrize("n",
+                         [0, 1, 4, None],
+                         ids=['0', '1', '4', 'None'])
+@pytest.mark.parametrize("increase",
+                         [True, False],
+                         ids=['True', 'False'])
+def test_vander(array, type, n, increase):
+    a_np = numpy.array(array, dtype=type)
+    a_dpnp = dpnp.array(a_np)
```
Comment (samir-nasibli) [inline tests/test_arraycreation.py:217]: Does this mean that input can be only `dparray`? 
If yes then add please needed condition for fallback on python level.

Also it is better give the same input for both array creation functions independently:
```suggestion
    a_np = numpy.array(array, dtype=type)
    a_dpnp = dpnp.array(array, dtype=type)
```

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 11]
```diff
@@ -127,6 +127,38 @@ void dpnp_ones_like_c(void* result, size_t size)
...
+        for (size_t i = 0; i < size_in; ++i) { result[i * N] = 1; }
+        for (size_t i = 1; i < N; ++i){
+            for (size_t j = 0; j < size_in; ++j){
+                result[j * N + i] = result[j * N + i - 1] * array_in[j];
+            }
+        }
+
+    } else {
+        int N_int = static_cast<int>(N);
+        int size_in_int = static_cast<int>(size_in);
```
Comment (samir-nasibli) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:152]: Are you sure, that you need this casts?

File: dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx [hunk 12]
```diff
@@ -311,6 +313,24 @@ cpdef dparray dpnp_triu(dparray m, int k):
     return result
 
 
+cpdef dparray dpnp_vander(dparray x1, int N, int increasing):
+    """ Convert string type names (dparray.dtype) to C enum DPNPFuncType """
```
Comment (samir-nasibli) [inline dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx:317]: Please, use `#` for comments.

Cython converts this to strings.

**Review by samir-nasibli — CHANGES_REQUESTED** [review]

File: dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx (hunk 12 above)
Comment (samir-nasibli) [reply dpnp/dpnp_algo/dpnp_algo_arraycreation.pyx:317]: it's better remove them at all.

File: tests/test_arraycreation.py [hunk 13]
```diff
@@ -193,3 +193,29 @@ def test_triu_size_null(k):
...
+                         ids=['float64', 'float32', 'int64', 'int32', 'bool', 'complex128'])
+@pytest.mark.parametrize("n",
+                         [0, 1, 4, None],
+                         ids=['0', '1', '4', 'None'])
+@pytest.mark.parametrize("increase",
+                         [True, False],
+                         ids=['True', 'False'])
+def test_vander(array, type, n, increase):
+    a_np = numpy.array(array, dtype=type)
+    a_dpnp = dpnp.array(a_np,  dtype=type)
```
Comment (densmirn) [inline tests/test_arraycreation.py:217]:
```suggestion
    a_dpnp = dpnp.array(a_np, dtype=type)
```

File: tests/test_arraycreation.py [hunk 14]
```diff
@@ -193,3 +193,29 @@ def test_triu_size_null(k):
...
+@pytest.mark.parametrize("array",
+                         [[1, 2, 3, 4],
+                             [],
+                             [0, 3, 5]],
+                         ids=['[1, 2, 3, 4]',
+                              '[]',
+                              '[0, 3, 5]'])
+@pytest.mark.parametrize("type",
+                         [numpy.float64, numpy.float32, numpy.int64,
+                             numpy.int32, numpy.bool, numpy.complex128],
```
Comment (densmirn) [inline tests/test_arraycreation.py:207]:
```suggestion
                          numpy.int32, numpy.bool, numpy.complex128],
```

File: tests/test_arraycreation.py [hunk 15]
```diff
@@ -193,3 +193,29 @@ def test_triu_size_null(k):
     expected = numpy.triu(a, k)
     result = dpnp.triu(ia, k)
     numpy.testing.assert_array_equal(expected, result)
+
+
+@pytest.mark.parametrize("array",
+                         [[1, 2, 3, 4],
+                             [],
+                             [0, 3, 5]],
```
Comment (densmirn) [inline tests/test_arraycreation.py:201]:
```suggestion
                          [],
                          [0, 3, 5]],
```

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp (hunk 11 above)
Comment (densmirn) [reply dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:157]: I would use `size_t` instead of `int` in loops below and avoid using extra variables.

**Review by densmirn — APPROVED** [review]

File: dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp [hunk 16]
```diff
@@ -127,6 +127,62 @@ void dpnp_ones_like_c(void* result, size_t size)
...
+        return;
+
+    const _DataType_input* array_in = reinterpret_cast<const _DataType_input*>(array1_in);
+    _DataType_output* result = reinterpret_cast<_DataType_output*>(result1);
+
+    if (N == 1)
+    {
+        for (size_t i = 0; i < size_in; ++i)
+        {
+            result[i] = 1;
```
Comment (densmirn) [inline dpnp/backend/kernels/dpnp_krnl_arraycreation.cpp:146]: You could execute function `dpnp_ones_c` here. Isn't it?

File: tests/test_arraycreation.py [hunk 17]
```diff
@@ -193,3 +193,29 @@ def test_triu_size_null(k):
...
+                         ids=['float64', 'float32', 'int64', 'int32', 'bool', 'complex128'])
+@pytest.mark.parametrize("n",
+                         [0, 1, 4, None],
+                         ids=['0', '1', '4', 'None'])
+@pytest.mark.parametrize("increase",
+                         [True, False],
+                         ids=['True', 'False'])
+def test_vander(array, type, n, increase):
+    a_np = numpy.array(array, dtype=type)
+    a_dpnp = dpnp.array(a_np, dtype=type)
```
Comment (samir-nasibli) [inline tests/test_arraycreation.py:217]: Again.
Please, give the same input for both array creation functions independently:
```suggestion
    a_dpnp = dpnp.array(array, dtype=type)
```

**Review by shssf — APPROVED** [review]

**Pull request merged**